.github/workflows/cleanup.yml

```diff
@@ -176,3 +189,99 @@ jobs:
+        with:
+          github-token: '${{ github.token }}'
+          retries: '3'
+          script: |+
+            var body, req;
+            const fs = require("fs");
+            const outcome = '${{ steps.cleanup_iam.outcome }}';
+            switch (outcome) {
+              case 'success':
+                req = fs.readFileSync(
+                  `${{ env.IAM_OUT_FILENAME }}`,
+                  { encoding: "utf8" }
+                );
+
+              body = `**\`Access on Demand\`** - 🟩 **\`IAM\`** request cleanup succeeded.
+
+            <details>
+            <summary>Details</summary>
+            Removed bindings in the request below.
+
+            \`\`\`
+            ${req}
+            \`\`\`
+            </details>`;
+                break;
+              case 'failure':
+                req = fs.readFileSync(
+                  `${{ env.IAM_OUT_FILENAME }}`,
+                  { encoding: "utf8" }
+                );
+                const error = fs.readFileSync(
+                  `${{ env.IAM_ERROR_FILENAME }}`,
+                  { encoding: "utf8" }
+                );
+                body = `**\`Access on Demand\`** - 🟥 **\`IAM\`** request cleanup failed.
+
+            <details>
+            <summary>Details</summary>
+            Failed to cleanup IAM polices of the resources in the request below.
+
+            \`\`\`
+            ${req}
+            \`\`\`
+
+            Error:
+            \`\`\`
+            ${error}
+            \`\`\`
+            </details>`;
+                break;
+              // step cancelled/skipped.
+              default:
+                // Do nothing.
+                break;
+            }
+
+            if (typeof body !== "undefined") {
+              await github.rest.issues.createComment({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                issue_number: ${{ github.event.pull_request.number }},
+                body: body,
+              });
+            }
+      # Separate step for request not found, as we cannot tell from step outcome
+      # whether the cleanup step is skipped because of request file not found or
+      # previous step failures.
+      - name: 'IAM Request Not Found Comment'
+        if: '${{ always() && hashFiles(''iam.yaml'') == '''' }}'
+        uses: 'actions/github-script@98814c53be79b1d30f795b907e553d8679345975' # ratchet:actions/github-script@v6
+        with:
+          github-token: '${{ github.token }}'
+          retries: '3'
+          script: |+
+            await github.rest.issues.createComment({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              issue_number: ${{ github.event.pull_request.number }},
+              body: `**\`Access on Demand\`** - 🟦 **\`IAM\`** request not found, skip cleanup.`,
+            });
+      - name: 'Tool Request Not Found Comment'
+        if: '${{ always() && hashFiles(''tool.yaml'') == '''' }}'
```

Comment:
Just curious what does always() && hashFiles(''tool.yaml'') == '''' }} indicates?

Reply:
If it always run if tool.yaml file exists, should it be '${{ always() && hashFiles(''tool.yaml'') != '''' }}', in addition, what's the difference between '${{ always() && hashFiles(''tool.yaml'') != '''' }}' and directly using '${{ always() && hashFiles(''tool.yaml'') != '''' }}'

Reply:
Sorry, I later edited my response as I made a mistake, it runs when file does not exist. Is there a typo in your question -> ```what's the difference between '${{ always() && hashFiles(''tool.yaml'') != '''' }}' and directly using '${{ always() && hashFiles(''tool.yaml'') != '''' }}'``` the former one is the same as the later one.

Reply:
I mean what's the difference between '${{ always() && hashFiles(''tool.yaml'') != '''' }}' and directly using '&hashFiles(''tool.yaml'') != '''' }}'

Reply:
steps in jobs by default has dependency, if any previous steps failed or skipped, it wont run unless with explicit condition like always().